src/TimeoutMigrationTool/Asp/AspTimeoutsSource.cs

```diff
@@ -0,0 +1,608 @@
+                foreach (var entitiesInTheSamePartition in seg.Results.GroupBy(x => x.PartitionKey))
+                {
+                    foreach (var timeoutDataEntity in entitiesInTheSamePartition)
+                    {
+                        // entries with Guid as partition key should never be modified. We are only interested in the query entities
+                        if (Guid.TryParse(timeoutDataEntity.PartitionKey.AsSpan(), out _))
+                        {
+                            continue;
+                        }
+
+                        if (hideByPartitionBatch.Count == 100)
+                        {
+                            await endpointTimeoutTable.ExecuteBatchAsync(hideByPartitionBatch);
+                            hideByPartitionBatch.Clear();
+                        }
+
+                        // we don't want to preserve the etag and we always want to win (is that true?)
+                        timeoutDataEntity.ETag = "*";
+                        // Fix the hiding part
+                        timeoutDataEntity.OwningTimeoutManager = $"{AspConstants.MigrationOngoingPrefix}{timeoutDataEntity.OwningTimeoutManager}";
+
+                        hideByPartitionBatch.Add(TableOperation.Merge(timeoutDataEntity));
+                    }
+
+                    if (hideByPartitionBatch.Count > 0)
+                    {
+                        await endpointTimeoutTable.ExecuteBatchAsync(hideByPartitionBatch);
+                    }
+
+                    hideByPartitionBatch.Clear();
+                }
+
+                // with batching we can at least efficiently insert
+                var migrationBatch = new TableBatchOperation();
+                foreach (var timeoutDataEntity in seg.Results)
+                {
+                    // entries with Guid as partition key should never be modified. We are only interested in the query entities
+                    if (Guid.TryParse(timeoutDataEntity.PartitionKey.AsSpan(), out _))
+                    {
+                        continue;
+                    }
+
+                    previousBatch = currentBatch;
+                    currentBatch = (totalNumberOfItems / batchSize) + 1;
+
+                    // there is no need to download the state from blob storage yet, we can do that when migrating the batches
+                    var migratedTimeoutDataEntity = new MigratedTimeoutDataEntity(timeoutDataEntity, currentBatch, AspConstants.MigrationOngoingPrefix);
+
+                    if (currentBatch != previousBatch)
+                    {
+                        batchNumberAndSizes.Add((previousBatch, numberOfItemsInBatch));
+                        numberOfItemsInBatch = 0;
+                    }
+
+                    // the batch can have max 100 items and max 4 MB of data
+                    // the partition key for all operations in the batch has to be the same
+                    if (currentBatch != previousBatch || migrationBatch.Count == 100)
```

Comment:
Shouldn't we check for the 4MB size too?

Reply:
See https://github.com/Particular/TimeoutMigrationTool/pull/172#discussion_r570014836